### DISCUSSION TITLE
Support of selecting collection's fields

### DISCUSSION THREAD

**Discussion by koltyakov** [pocketbase/pocketbase#354] [Q&A]
Hi there! I'm new to PocketBase yet I looks promising and nicely crafted tool. I've found one of the features missed and wondering are there any plans adding REST/OData's `select` capability to the API? To be able limiting responses with only requested fields to the client. It might be especially handy to use with `expand`s when only few fields are usually needed from within related entities. E.g. `?select=*,rel1.field1,rel1.field2`.

**Comment by ganigeorgiev** [comment]
This was partially discussed in https://github.com/pocketbase/pocketbase/issues/259.

This usually shouldn't be very hard to implement, but currently this is not planned since it may complicate the integration with the eventual hidden fields support in the future, so I'll have to think a little more on this before committing and adding it to the roadmap.
It would be a nice to have feature, but it is mostly an optimization and it shouldn't be a blocker for people to use the existing web APIs.

I'll revisit this and the hidden fields proposals once again in the near future, together with the planned framework internals refactoring and the new documentation (it will be somewhere after v0.8/0.9 release).

_btw. If you haven't already, you can check the current v1.0.0 roadmap here - https://github.com/orgs/pocketbase/projects/2._

**Reply by koltyakov** [reply]
Thank you @ganigeorgiev for the instant answer!

**Reply by sonovice** [reply]
> is mostly an optimization and it shouldn't be a blocker for people to use the existing web APIs.

In my case it is a show stopper since the returned data is hugely blown up in size (factor x20) due to verbosity. Instead of implementing it in the SQL query layer, it could simply be a filter on the already collected JSON data just before returning it to the client. Not sure whether you like this approach.

**Reply by sonovice** [reply]
> it could simply be a filter on the already collected JSON data just before returning it to the client

I am trying to do just that but am struggling on how to get access to the expanded data:
```go
app.OnRecordViewRequest("music_collections").Add(func(e *core.RecordViewEvent) error {
	foo := e.Record.Expand() // ? -> returns map[]
	fmt.Printf("%v\n", foo)
	return nil
})
```
My final goal is to remove the field `songs.lyrics` with `songs` being an expanded field of `music_collections`.
Any help on this would be highly appreciated.

**Reply by ganigeorgiev** [reply]
@sonovice The records enriching (including expand) is done at the last stage, right before writing the records to the response body.

If you want to eagerly enrich the records (aka. `e.Record.Expand()` to be populated), you can use [`apis.EnrichRecord`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.13.0/apis#EnrichRecord).

Please keep in mind that even if you modify the expanded records and unset the specific field value, the record fields will still be returned in the response.
If you want to provide a custom response from the hook, you have to stop its propagation with `hook.StopPropagation`:

```go
e.HttpContext.JSON(http.StatusOK, yourCustomResponseData)

return hook.StopPropagation
```

Alternatively to modifying the response, you can also make use of the new "view" collection type available with v0.13.0 and create a new read-only collection with subset of the existing fields.

**Reply by sonovice** [reply]
Thank you. I just checked the new "view" type and it seems to solve my problems. However, if I create a `songs_without_lyrics` view, would I still have to populate my mentioned `music_collections` record with each song again? Not sure how to use the expand feature any other way.

**Reply by ganigeorgiev** [reply]
I'm lacking details about your collections structure to fully understand what you are asking and what you are trying to do.

If you've created a new view `songs_without_lyrics` then this is treated as a new separate collection (even if it shares the same records as the original base collection).

The `relation` field has a reference to the collection shown in its field options (under the `options.collectionId` key). Once set it cannot be changed without manually updating the db collection entry. If you really want to do that, you can do it with a migration, something like (_I haven't tested it and I'm still not sure that I understand your db structure, so keep a backup of your data before running it_):

```go
package migrations

import (
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/daos"
	m "github.com/pocketbase/pocketbase/migrations"
	"github.com/pocketbase/pocketbase/models/schema"
)

func init() {
	m.Register(func(db dbx.Builder) error { // up
		dao := daos.New(db)

		songsWithoutLyrics, err := dao.FindCollectionByNameOrId("songs_without_lyrics")
		if err != nil {
			return err
		}

		musicCollections, err := dao.FindCollectionByNameOrId("music_collections")
		if err != nil {
			return err
		}

		field := musicCollections.Schema.GetFieldByName("songs")
		options := field.Options.(*schema.RelationOptions)
		options.CollectionId = songsWithoutLyrics.Id
		// change something else...

		// optionally you can even replace its reference (the replacement is done via the field id)
		musicCollections.Schema.AddField(field)

		// persist changes
		return dao.SaveCollection(musicCollections)
	}, func(db dbx.Builder) error { // down
		dao := daos.New(db)

		songs, err := dao.FindCollectionByNameOrId("songs")
		if err != nil {
			return err
		}

		musicCollections, err := dao.FindCollectionByNameOrId("music_collections")
		if err != nil {
			return err
		}

		field := musicCollections.Schema.GetFieldByName("songs")
		options := field.Options.(*schema.RelationOptions)
		options.CollectionId = songs.Id
		// change something else...

		// optionally you can even replace its reference (the replacement is done via the field id)
		musicCollections.Schema.AddField(field)

		// persist changes
		return dao.SaveCollection(musicCollections)
	})
}
```

**Reply by sonovice** [reply]
Thanks, I really appreciate your help, especially your migration script. However, at this point it feels like I am trying to massage my database structure to a very specific use case (selecting fields on an expanded relation). A custom endpoint might be the saner option here with a raw SQL query.

**Comment by ganigeorgiev** [comment]
Just for info in case someone is searching for this - support for `?fields=field1,field2` query parameter to limit the returned API fields was added in the [`develop`](https://github.com/pocketbase/pocketbase/tree/develop) branch and will be available with the next PocketBase v0.16.0 release (there are no ETAs yet).
